chain-registry/src/formatter.rs

```diff
@@ -0,0 +1,233 @@
+//! from an (in)complete GRPC URL.
+use crate::error::RegistryError;
+use http::uri::Scheme;
+use http::Uri;
+use std::str::FromStr;
+
+use tendermint_rpc::Url;
+
+/// `UriFormatter` contains the basic expectations to parse a valid URL from a `&str`.
+pub trait UriFormatter {
+    /// Expected output format of the formatter.
+    type OutputFormat;
+
+    /// Attempts to parse the given input as a OutputFormat. If the parsed URL
+    /// is not complete, this method attempts to fill in the necessary missing
+    /// pieces. Returns a RegistryError if this attempt fails.
+    ///
+    /// # Arguments
+    ///
+    /// * `url` - A string slice that holds the url which should be formatted
+    fn parse_or_build_address(url: &str) -> Result<Self::OutputFormat, RegistryError>;
+}
+
+/// `SimpleWebSocketFormatter` contains methods to parse a valid WebSocket URL from an RPC URL.
+pub struct SimpleWebSocketFormatter;
+
+/// `SimpleGrpcFormatter` contains methods to parse or build a valid `http::Uri` from an (in)complete GRPC URL.
+pub struct SimpleGrpcFormatter;
+
+impl UriFormatter for SimpleWebSocketFormatter {
+    type OutputFormat = Url;
+
+    /// Format a WebSocket URL from an RPC URL and return a `tendermint_rpc::Url`.
+    fn parse_or_build_address(rpc_address: &str) -> Result<Self::OutputFormat, RegistryError> {
+        let uri = rpc_address
+            .parse::<Uri>()
+            .map_err(|e| RegistryError::uri_parse_error(rpc_address.to_string(), e))?;
+
+        let uri_websocket = Uri::builder()
+            .scheme(if uri.scheme().unwrap_or(&Scheme::HTTP) == &Scheme::HTTPS {
+                "wss"
+            } else {
+                "ws"
+            })
+            .authority(
+                uri.authority()
+                    .ok_or_else(|| {
+                        RegistryError::rpc_url_without_authority(rpc_address.to_string())
+                    })?
+                    .clone(),
+            )
+            .path_and_query("/websocket")
+            .build();
+
+        match uri_websocket {
+            Ok(uri_websocket) => Ok(Url::from_str(uri_websocket.to_string().as_str()).map_err(
+                |e| RegistryError::tendermint_url_parse_error(rpc_address.to_string(), e),
+            )?),
+            Err(e) => Err(RegistryError::unable_to_build_websocket_endpoint(
+                rpc_address.to_string(),
+                e,
+            )),
+        }
+    }
+}
+
+/// Builds a valid `http::Uri` from a gRPC address.
+impl UriFormatter for SimpleGrpcFormatter {
+    type OutputFormat = Uri;
+
+    fn parse_or_build_address(input: &str) -> Result<Self::OutputFormat, RegistryError> {
+        // Remove the last character if it is a '/'
+        let input = match input.ends_with('/') {
+            false => input,
+            true => {
+                let mut chars = input.chars();
+                chars.next_back();
+                chars.as_str()
+            }
+        };
```

Comment:
It's more idiomatic to use an `if` statement on boolean checks:
```suggestion
        let input = if input.ends_with('/') {
            let mut chars = input.chars()
            chars.next_back();
            chars.as_str()
        } else {
            input
        };
```